pkg/controller/cloudfront/distribution/setup.go

```diff
@@ -0,0 +1,182 @@
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/cloudfront/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// TODO: isn't this defined as an API constant somewhere in aws-sdk-go? Generated zz_enums.go seems not to contain it either
+const stateDeployed = "Deployed"
+
+// SetupDistribution adds a controller that reconciles Distribution.
+func SetupDistribution(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(svcapitypes.DistributionGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&svcapitypes.Distribution{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.DistributionGroupVersionKind),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithExternalConnecter(&connector{
+				kube: mgr.GetClient(),
+				opts: []option{
+					func(e *external) {
+						e.preCreate = preCreate
+						e.postCreate = postCreate
+						e.lateInitialize = lateInitialize
+						e.preObserve = preObserve
+						e.postObserve = postObserve
+						e.isUpToDate = isUpToDate
+						e.preUpdate = preUpdate
+						e.preDelete = preDelete
+					},
+				},
+			}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.Distribution, cdi *svcsdk.CreateDistributionInput) error {
+	cdi.DistributionConfig.CallerReference = awsclients.String(string(cr.UID))
+	cdi.DistributionConfig.Origins.Quantity =
+		awsclients.PInt64Value(len(cr.Spec.ForProvider.DistributionConfig.Origins.Items))
+
+	return nil
+}
+
+func postCreate(_ context.Context, cr *svcapitypes.Distribution, cdo *svcsdk.CreateDistributionOutput,
+	ec managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+
+	meta.SetExternalName(cr, awsclients.StringValue(cdo.Distribution.Id))
+	ec.ExternalNameAssigned = true
+
+	return ec, nil
+}
+
+func preObserve(_ context.Context, cr *svcapitypes.Distribution, gdi *svcsdk.GetDistributionInput) error {
+	gdi.Id = awsclients.String(meta.GetExternalName(cr))
+
```

Comment:
Very old habit from Deitel & Deitel programming books. The idea is to have all declarations, control blocks, etc. stand out for "ease of reading". But this's not aligned with go conventions. The more space style is easier to my eyes and because `gofmt -s` does not remove them (except for repeated empty lines), I have liberally used them but agreed, I will be more careful the repo conventions.